napari/components/_layer_slicer.py

```diff
@@ -130,6 +139,24 @@ def slice_layers_async(
         wait for any existing tasks that include that layer AND another layer,
         In other words, it will only cancel if the new task will replace the
         slices of all the layers in the pending task.
+
+        Parameters
+        ----------
+        layers: iterable of layers
+            The layers to slice.
+        dims: Dims
+            The dimensions values associated with the view to be sliced.
+        _refresh_sync: bool
+            True if when forcing synchronous slicing, `refresh` should be used
+            instead of `_slice_dims`. False otherwise. The leading underscore
+            indicates that this is a temporary parameter that will be removed
+            after old-style slicing is removed.
```

Comment:
I'm struggling to follow this flow. Can you elaborate a bit on the `refresh` vs `_slice_dims` purpose? Does this basically mean that as long as there is one non-async layer, everythign will be forced to sync?

Reply:
Yeah, this is some ugliness I don't like. The main purpose is to force slicing to occur even when the slicing input that came from `Dims` didn't change. And it's to support existing use of `Layer.refresh` in the code (e.g. in Layer initialization and in many other places).

On `main`, `Layer._slice_dims` is mostly `Layer.refresh` with the following differences.

- If the slice input passed to `_slice_dims` is the same as the current one, it will return early and do nothing else.
- If `dims.ndisplay` has changed, it will emit the `_ndisplay` event.
- It also calls `Layer._set_editable` after `refresh`, which should probably be renamed `reset_editable`.

The early return is the main thing to worry about because it can skip slicing.

This PR causes `ViewerModel._update_layers` to do slicing through `LayerSlicer`, but it also causes the refresh in `Layer._update_draw` to go through `LayerSlicer` since that's how we currently handle FOV changes in 2D multi-scal images (I also see that as a type of slicing).

In the new slicing path (of executing requests), we don't really distinguish between these two anymore - everything is more like `_slice_dims` because we make a slice request from the Layer and Dims then may execute it or may cancel it. We may want some ways to mirror the early return for performance reasons (e.g. by return `None` in `Layer._make_slice_request`), but I was leaving those optimizations for later, especially when the async path can achieve some performance gain in some cases by being lazy.

We can remove this parameter when all layers support the request/response style of slicing. I was actually playing with a PR that does this somewhat trivially, and may be one way to avoid some of this ugliness.

> Does this basically mean that as long as there is one non-async layer, everythign will be forced to sync?

No. That's something I had in the prototype, but we decided to allow async layers to be sliced concurrently with sync layers (when async slicing is on).